torchdynamo/eval_frame.py

```diff
@@ -256,50 +273,83 @@ def dynamo_normalization_capturing_compiler(
     # TODO(voz): Handle kwargs properly?
     flat_args, in_spec = pytree.tree_flatten(args)
-    for arg in flat_args:
-        flat_input_types.append(arg.__class__)
+    dynamo_flat_args, dynamo_in_spec = pytree.tree_flatten(compiler_captured_inputs)
+
+    if in_spec != dynamo_in_spec:
+        if len(dynamo_flat_args) > 0:
+            assert_can_rewrite_spec(
+                dynamo_flat_args, flat_args, in_spec, dynamo_in_spec, args
+            )
 
-    _, out_spec = pytree.tree_flatten(result)
+    flat_results_traced, out_spec_traced = pytree.tree_flatten(result_traced)
 
-    assert len(flat_input_types) == len(
-        out_sig.parameters
-    ), "Flattened inputs length must match out signature parameter lengths."
+    result_export = graph.forward(*flat_args)
+    flat_results_export, out_spec_export = pytree.tree_flatten(result_export)
 
-    for idx in range(len(out_sig.parameters)):
-        sig_type = signature_types[idx]
-        in_type = flat_input_types[idx]
-        assert (
-            sig_type == in_type
-        ), "Export produced a graph with mismatched type signature {sig_type} vs expected {in_type} for arg {idx}"
+    flat_inputs_to_exported_graph, flat_inputs_spec = pytree.tree_flatten(
+        compiler_captured_inputs
+    )
+
+    if out_spec_export != out_spec_traced:
+        flat_both = flat_inputs_to_exported_graph + flat_results_export
+        assert_can_rewrite_spec(
+            flat_both,
+            flat_results_traced,
+            out_spec_traced,
+            out_spec_export,
+            result_traced,
+        )
 
-    # TODO(voz): A major feature gap here, atm, is that we return a graph with a flat_args signature.
+    # NOTE: This is a lengthy outline of major UX gaps, but the primary goal is to produce
+    # a graph with the same exact input and output signatures as the original traced callable.
+
+    # 1) TODO(voz): call _codegen on the graph to rewrite the graphs input and output
+    # Blocked on bugs in fx. https://github.com/pytorch/pytorch/pull/81177
+    # graph.graph._codegen = _PyTreeCodeGen(
+    # _PyTreeInfo(
+    #         [f"orig_arg_{i}" for i in range(len(inputs))],
+    #         in_spec,
+    #         out_spec_traced,
+    #     )
+    # )
+    # graph.recompile()
+
+    # 2) NOTE: This may not be sufficient! In the event that we have args that pass around the graph,
+    # and are elided by dynamo, the reconstruct code above places them into the output correctly.
+    # However, _PyTreeCodeGen does not do that. If we are to ensure our users have a good UX, we
+    # need to do the codegen reconstruction for them.
+    # This gives us a few major paths:
+    #  A) Expose a util function that takes the output of this (graph, specs, etc)
+    # and lets users call it with regular function args, and that function handles the input remapping.
+    #  B) We write out own version of _PyTreeCodeGen that generates the remapping and makes it into the graph.
```

Comment:
B, so far, I think, is the best option here, because if we want to offer the user convenience functions short of rewriting (codegen + recompile) the graph (like a util or wrapper func), we would have to end up calling forward twice - once to capture the 'eager' args, and once on the exported graph, so that we can reconstruct. Codegen + recompile avoids this problem.

Reply:
Thanks, did not know about `Transformer` - this looks perfect. Order should be very easy to extract from the functions already in this PR.